IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,220 @@
+        MSIDCredentialCacheItem *credential = [self storedCredentialForKey:credentialKey];
+        if (credential)
+        {
+            return @[credential];
+        }
+        
+        return nil;
+    }
+    
+    NSMutableArray *subPredicates = [[NSMutableArray alloc] init];
+    
+    /*
+     Refresh token either has family Id as 1 or nil.
+     For app refresh token, we pass family id as nil which matches both FRT and RT instead of matching just RT
+     For clients that are not part of family, family is saved as empty string and added to refresh token query.
+     */
+    if ([NSString msidIsStringNilOrBlank:key.familyId])
+    {
+        key.familyId = nil;
+    }
+
+    [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.familyId == %@", key.familyId]];
+    if (key.clientId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.clientId == %@", key.clientId]];
+    if (key.environment)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.environment == %@", key.environment]];
+    if (key.homeAccountId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.homeAccountId == %@", key.homeAccountId]];
+    if (key.credentialType)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.credentialType == %d", key.credentialType]];
+    if (key.realm)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.realm == %@", key.realm]];
+    /*
+     key.target is not added as target can be subset , intersect , superset or exact string match and is matched later in the code.
+     */
+    
+    // Combine all sub-predicates with AND:
+    NSPredicate *matchAttributes = [NSCompoundPredicate andPredicateWithSubpredicates:subPredicates];
+    
+    NSArray *storedCredentials = [self allValues];
+    NSArray *filteredCredentials = [storedCredentials filteredArrayUsingPredicate:matchAttributes];
+    return filteredCredentials;
+}
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(NSError * __autoreleasing *)error
+{
+    if (!(self = [self init]))
+    {
+        return nil;
+    }
+    
+    for (NSString *credentialKey in json)
+    {
+        NSDictionary *rtDict = [json objectForKey:credentialKey];
+        
+        if (rtDict)
+        {
+            MSIDCredentialCacheItem *credential = [[MSIDCredentialCacheItem alloc] initWithJSONDictionary:rtDict error:error];
+            
+            if (credential)
+            {
+                [self storeCredential:credential forKey:credentialKey];
+            }
+        }
+    }
+    
+    return self;
+}
+
+- (NSDictionary *)jsonDictionary
+{
+    NSArray *credentialKeys = [self allKeys];
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    
+    for (NSString *credentialKey in credentialKeys)
+    {
+        MSIDCredentialCacheItem *credential = [self storedCredentialForKey:credentialKey];
```

Comment:
should we synchronize entire operation instead of synchronizing each read?

Reply:
Here we don't need to synchronize the entire operation as it will effect performance. Just synchronizing the read will be faster.

Reply:
Is it guaranteed that nobody changes dictionary in between?

Reply:
Changes to dictionary will happen if we add/update or remove an item and that should always be followed by a new call to jsonDictionary to update the new serialized blob.